Frameworks/Foundation/NSOperation.mm

```diff
@@ -131,139 +128,109 @@ - (void)setThreadPriority:(double)priority {
  @Status Interoperable
 */
 - (void)setCompletionBlock:(void (^)(void))block {
-    id oldBlock = priv->completionBlock;
-    priv->completionBlock = [block copy];
-    [oldBlock release];
+    pthread_mutex_lock(&_completionBlockLock);
+    [_completionBlock autorelease];
+    _completionBlock = [block copy];
+    pthread_mutex_unlock(&_completionBlockLock);
 }
 
 /**
  @Status Interoperable
 */
 - (void (^)(void))completionBlock {
-    return priv->completionBlock;
-}
-
-/**
- @Status Interoperable
-*/
-- (BOOL)isReady {
-    //  Note, check dependencies when we get them
-    int count = [priv->dependencies count];
-
-    for (int i = 0; i < count; i++) {
-        id op = [priv->dependencies objectAtIndex:i];
-
-        if (![op isFinished]) {
-            return NO;
-        }
-    }
-    return YES;
+    pthread_mutex_lock(&_completionBlockLock);
+    id ret = [[_completionBlock retain] autorelease];
+    pthread_mutex_unlock(&_completionBlockLock);
+    return ret;
 }
 
 /**
  @Status Interoperable
 */
 - (BOOL)isCancelled {
-    return priv->cancelled != 0;
+    return _cancelled;
 }
 
 /**
  @Status Interoperable
 */
 - (BOOL)isFinished {
-    return priv->finished != 0;
+    return _finished;
 }
 
 /**
  @Status Interoperable
 */
 - (BOOL)isExecuting {
-    return priv->executing != 0;
+    return _executing;
 }
 
 /**
  @Status Interoperable
 */
 - (void)start {
-    if (!priv->executing && !priv->finished) {
-        bool execute = false;
-
-        pthread_mutex_lock(&priv->finishLock);
-        if (!priv->cancelled) {
-            [self willChangeValueForKey:@"isExecuting"];
-            priv->executing = 1;
-            [self didChangeValueForKey:@"isExecuting"];
-            execute = true;
-        }
-        pthread_mutex_unlock(&priv->finishLock);
-        if (execute) {
-            [self main];
-        }
-
-        if (execute) {
-            [self willChangeValueForKey:@"isExecuting"];
-        }
+    if (_finished) {
+        return;
+    }
 
-        [self _setFinished:YES
-            andPerformUnderLock:^{
-                if (execute) {
-                    priv->executing = 0;
-                    [self didChangeValueForKey:@"isExecuting"];
-                }
-            }];
+    THROW_NS_IF(E_INVALIDARG, (_executing || ![self isReady]));
+    
+    pthread_mutex_lock(&_finishLock);
+    BOOL shouldExecute = !_cancelled; // Note: in the cancelled case, [self main] is not called but isFinished will still be observable
+    if (shouldExecute) {
+        [self willChangeValueForKey:@"isExecuting"];
+        _executing = YES;
+        [self didChangeValueForKey:@"isExecuting"];
+    }
+    pthread_mutex_unlock(&_finishLock);
+    
+    if (shouldExecute) {
+        NSAutoreleasePool* pool = [NSAutoreleasePool new];
+        [self main];
```

Comment:
According to the preceding sentence in the documentation, this is the correct behavior. " In OS X v10.6 and later, exceptions are allowed to propagate beyond this [start] method. You should never allow exceptions to propagate out of your main method."

---
In reply to: [69355325](https://github.com/Microsoft/WinObjC/pull/602#discussion_r69355325) [](ancestors = 69355325)